db/hybrid_logical_vector.go

```diff
@@ -230,52 +230,40 @@ func (hlv *HybridLogicalVector) DominatesSource(version Version) bool {
 
 // AddVersion adds newVersion as the current version to the in memory representation of the HLV.
 func (hlv *HybridLogicalVector) AddVersion(newVersion Version) error {
-	var newVersionCAS uint64
-	hlvVersionCAS := hlv.Version
-	if newVersion.Value != expandMacroCASValueUint64 {
-		newVersionCAS = newVersion.Value
-	}
+
 	// check if this is the first time we're adding a source - version pair
 	if hlv.SourceID == "" {
 		hlv.Version = newVersion.Value
 		hlv.SourceID = newVersion.SourceID
-		hlv.InvalidateMV()
 		return nil
 	}
-	// if new entry has the same source we simple just update the version
+
+	// If the new version is older than an existing version for the same source, return error
+	existingValueForSource, found := hlv.GetValue(newVersion.SourceID)
+	if found && existingValueForSource > newVersion.Value {
+		return fmt.Errorf("attempting to add new version vector entry with a value that is less than the existing value for the same source. New version: %v, Existing HLV: %v", newVersion, hlv)
+	}
+
+	// Move existing mv to pv before adding the new version
+	hlv.InvalidateMV()
```

Comment:
The comment in `InvalidateMV` is wrong:

```
// InvalidateMV will move all merge versions to PV, except merge version entries that share a source with cv
```

I think this should be:
```
// InvalidateMV will move all merge versions to previous versions. This function should be called before placing current version in previous versions, since a cv will override an mv.
```

Reply:
I think the first version is actually accurate? (merge versions for the source cv will not be moved to pv)

Reply:
I misread this function, looks good.